test/lit/passes/rse-eh.wast

```diff
@@ -0,0 +1,275 @@
+  ;; CHECK-NEXT:      (try_table (catch_all_ref $catch_all_ref1)
+  ;; CHECK-NEXT:       (throw $e-i32
+  ;; CHECK-NEXT:        (i32.const 0)
+  ;; CHECK-NEXT:       )
+  ;; CHECK-NEXT:      )
+  ;; CHECK-NEXT:     )
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:    (local.set $x
+  ;; CHECK-NEXT:     (i32.const 1)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:    (throw_ref
+  ;; CHECK-NEXT:     (local.get $exn)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (i32.const 1)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $nested-try_table1
+    (local $x i32)
+    (local $exn exnref)
+    (block $catch_all0
+      (try_table (catch_all $catch_all0)
+        (local.set $exn
+          (block $catch_all_ref1 (result exnref)
+            (try_table (catch_all_ref $catch_all_ref1)
+              (throw $e-i32 (i32.const 0))
+            )
+          )
+        )
+        (local.set $x (i32.const 1))
+        (throw_ref (local.get $exn))
+      )
+    )
+    ;; The exception is caught by the inner catch_all_ref, which runs the
+    ;; local.set, so this should be dropped
+    (local.set $x (i32.const 1))
+  )
+
+  ;; CHECK:      (func $nested-try_table2 (type $0)
+  ;; CHECK-NEXT:  (local $x i32)
+  ;; CHECK-NEXT:  (local $exn exnref)
+  ;; CHECK-NEXT:  (local $pair (i32 exnref))
+  ;; CHECK-NEXT:  (block $catch_all0
+  ;; CHECK-NEXT:   (try_table (catch_all $catch_all0)
+  ;; CHECK-NEXT:    (local.set $pair
+  ;; CHECK-NEXT:     (block $catch1 (type $1) (result i32 exnref)
+  ;; CHECK-NEXT:      (try_table (catch_ref $e-i32 $catch1)
+  ;; CHECK-NEXT:       (throw $e-i32
+  ;; CHECK-NEXT:        (i32.const 0)
+  ;; CHECK-NEXT:       )
+  ;; CHECK-NEXT:      )
+  ;; CHECK-NEXT:     )
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:    (local.set $exn
+  ;; CHECK-NEXT:     (tuple.extract 2 1
+  ;; CHECK-NEXT:      (local.get $pair)
+  ;; CHECK-NEXT:     )
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:    (local.set $x
+  ;; CHECK-NEXT:     (i32.const 1)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:    (throw_ref
+  ;; CHECK-NEXT:     (local.get $exn)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (local.set $x
+  ;; CHECK-NEXT:   (i32.const 1)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $nested-try_table2
+    (local $x i32)
+    (local $exn exnref)
+    (local $pair (i32 exnref))
+    (block $catch_all0
+      (try_table (catch_all $catch_all0)
+        (local.set $pair
+          (block $catch1 (result i32 exnref)
+            (try_table (catch_ref $e-i32 $catch1)
+              (throw $e-i32 (i32.const 0))
+            )
+          )
+        )
+        (local.set $exn
+          (tuple.extract 2 1 (local.get $pair))
+        )
+        (local.set $x (i32.const 1))
+        (throw_ref (local.get $exn))
+      )
+    )
+    ;; Unlike nested-try_table1, the exception may not be caught by the inner
+    ;; catch, so the local.set may not run. So this should NOT be dropped.
```

Comment:
This is because the CFG builder does not consider tags when creating control flow edges, right? In reality the exception would always be caught here.

Reply:
Yeah, good point. I'll make a copy of this which throws with a different tag, and put a TODO on this one.

Reply:
Sorry, it shouldn't be a different tag, in which case it is always not gonna be caught. I'll replace the `throw` with a `call`, which we don't know what it will throw.